### PR TITLE
WIP: Ship graphs from client to scheduler with pickle

### PR DESCRIPTION
**Pull request by mrocklin**
This is just a proof of concept right now (many things fail) but this shows that we can ship graphs directly from the client to the scheduler without dealing with any pack/unpack things.  

We pickle the entire graph, ship it to the scheduler, where we then do all of the graph manipulation stuff that we used to do on the client.

cc @rjzamora @madsbk @ian-r-rose

### CODE REVIEW

**Comment by mrocklin** [comment]
So far I am surprised with how far I was able to get with relatively little lines of code changed.

There are, I'm aware, a whole army of monsters waiting for me.  Still though, this feels pretty ok so far.

**Comment by mrocklin** [comment]
Dask PR here: https://github.com/dask/distributed/pull/6028

**Comment by mrocklin** [comment]
OK, here is a class of tests that we're going to have problems with:

```python
    @gen_cluster(client=True)
    async def test_robust_unserializable(c, s, a, b):
        class Foo:
            def __getstate__(self):
                raise MyException()
    
        with pytest.raises(MyException):
>           future = c.submit(identity, Foo())
```

We're genuinely opening up the scheduler to more client issues.  I _think_ that the rewards outweigh the risks here, but there's definitely a regression here in terms of hardening.

**Comment by mrocklin** [comment]
OK, I spent a few hours hunting down a few of the test failures.  There are a few effects of moving the graph handling onto the scheduler which affect local behavior.  There are some things that we just get later in the process, and so they affect when users get negative feedback when they've done something wrong.  In practice these changes are small and, I think, not super likely to annoy people.  They do exist though.

As an example, if someone has a future from another client, or a cancelled future, we don't actually learn about it until after the graph is constructed and things are resolved on the client.  Whereas previously they would learn about this failure on a `submit` call, they now learn about it on a `result` call.  This is a degredation of experience, but not huge.

There are a few other things I can clean up (there is a long tail here) but in principle I think that the heavy parts of this are done.  I would welcome feedback on if this is a direction that we want to pursue.  

## TODO on this PR

- Fixup failing tests
- Merge `update_graph`/`update_graph_hlg`
- Jointly merge in the dask/dask PR

## Future Work

- Merge TaskGroups with Layers
- See how best we can simplify existing layers logic
- Remove cull, and replace with some nicer create_graph method

**Review by ian-r-rose — COMMENTED** [review]
Here's some early thoughts, I'm still mulling this over. Correctly forwarding annotations from layers to tasks (especially in the presence of any optimizations) still feels fragile, though the approach you've taken here makes sense.

Pickling certainly seems like a win from the client perspective, as well as for the implementer of HLG layers. It certainly does transfer some pain to the scheduler (which ultimately affects fewer maintainers, I suppose)

File: distributed/client.py [hunk 1]
```diff
@@ -2996,13 +3019,20 @@ def get(
         --------
         Client.compute : Compute asynchronous collections
         """
+        if isinstance(workers, (str, int)):
+            workers = [workers]
+        if allow_other_workers:
+            loose_restrictions = list(dsk)
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/client.py:3025]: Good point on `list(dsk)`.  And good idea with loose_restrictions.  I think that this is a case where I was going for an incremental change where really a revoluationary change (small revolution) would be easier and better for all involved.

File: distributed/scheduler.py [hunk 2]
```diff
@@ -4644,10 +4653,99 @@ def update_graph_hlg(
...
+                )
+                ts._exception = exc
+                ts._exception_text = text
+                ts._exception_blame = ts
+                parent.tasks[key] = ts
+
+            self.client_desires_keys(keys=keys, client=client)
+            return
+
+        dsk = dict(graph)
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/scheduler.py:4674]: Ideally low level graph optimization mostly goes away I think.  But in principle I agree that it can be done here.  We'll want to be mindful of annotations, but we would need to do that anyway.

I'm hopeful though that high level graph optimization removes the need here.

File: distributed/tests/test_client.py [hunk 3]
```diff
@@ -6686,7 +6695,6 @@ async def test_annotations_workers(c, s, a, b):
     with dask.config.set(optimization__fuse__active=False):
         x = await x.persist()
 
-    assert all({"workers": (a.address,)} == ts.annotations for ts in s.tasks.values())
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/tests/test_client.py:6689]: This isn't as clear to me that it's desired.  I don't mind pulling these out personally.  It's an open design choice I think.

File: distributed/scheduler.py [hunk 4]
```diff
@@ -4629,8 +4637,9 @@ async def add_nanny(self, comm):
 
     def update_graph_hlg(
         self,
-        client=None,
-        hlg=None,
+        client: str,
+        graph_header: bytes,
+        graph_frames: list[bytes],
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/scheduler.py:4642]: Woot.  thanks

File: distributed/scheduler.py [hunk 5]
```diff
@@ -4644,10 +4653,99 @@ def update_graph_hlg(
...
-        dsk = unpacked_graph["dsk"]
-        dependencies = unpacked_graph["deps"]
-        annotations = unpacked_graph["annotations"]
+        try:
+            graph: HighLevelGraph = pickle.loads(graph_header, buffers=graph_frames)
+        except Exception as e:
+            text = str(e)
+            exc = pickle.dumps(e)
+            parent: SchedulerState = cast(SchedulerState, self)
+            for key in keys:
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/scheduler.py:4682]: I've removed the default argument

File: distributed/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/client.py]: I've moved this logic up to the scheduler.

**Comment by mrocklin** [comment]
Update here.  Lots of things are working.  Some concerns;

1. We're no longer benefitting from custom Dask serialization of objects in the graph, like h5py files
2. Locks/Queues/Variables/Semaphores need special attention, which is fine, but things are proving a little tricky in inproc:// situations because we only serialize/deserialize once.

**Comment by rjzamora** [comment]
>Dask PR here: #6028

I assume this was meant to link [dask#8864](https://github.com/dask/dask/pull/8864) - (Just commenting to make that PR a bit easier to find)

File: distributed/scheduler.py [hunk 6]
```diff
@@ -4648,25 +4659,154 @@ async def add_nanny(self, comm):
...
+        dsk = dsk2
+
+        # - Add in deps for any tasks that depend on futures
+        for k, futures in fut_deps.items():
+            dependencies[k].update(f.key for f in futures)
+
+        pre_stringify = set(dsk)
+        dsk = {stringify(k): stringify(v, exclusive=graph) for k, v in dsk.items()}
+
+        def process(x, keys=dsk, string_keys=pre_stringify):
```
Comment (rjzamora) [inline distributed/scheduler.py:4750]: It sure would be nice if we didn't need to worry about stringifying keys anymore. Is the historical reason for doing this related to lists and tuples being indistinguishable in msgpack, or are there other reasons?

**Review by rjzamora — COMMENTED** [review]
Thanks again for pushing on this @mrocklin. My `distributed` knowledge is a bit limited, so I only have a few questions at this point.

>We're no longer benefitting from custom Dask serialization of objects in the graph, like h5py files

Right - This was indeed one of the reasons [dask#8672](https://github.com/dask/dask/pull/8672) was attempting to preserve the option to send the graph to the scheduler without using pickle (for now).  We would have eventually run into the same challenges in that case anyway, but I was hoping that would give us enough space to explore "custom"-serialization features in pickle. (Disclaimer: I am not at all confident that dask#8672 is a good approach - Just sharing some relevant motivation).

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -75,6 +75,10 @@ jobs:
         if: ${{ matrix.os == 'windows-latest' && matrix.python-version == '3.9' }}
         run: mamba uninstall ipython
 
+      - name: Install dask branch
+        shell: bash -l {0}
+        run: python -m pip install --no-deps git+https://github.com/mrocklin/dask@nuke-hlg
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply .github/workflows/tests.yaml:80]: Yeah, I think that there was some errant reduce function somewhere

File: distributed/client.py [hunk 8]
```diff
@@ -193,32 +192,38 @@ class Future(WrappedKey):
...
-            )
-
-        if state is not None:
+        if client is None and _current_client.get() is not False:
             try:
-                handler = self.client._state_handlers[state]
-            except KeyError:
+                client = Client.current()
+            except ValueError:
                 pass
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/client.py:199]: Correct.  In particular, Futures get pickled on the client and then unpickled on the scheduler, which doesn't have a client lying around.  We then quickly unpack these futures into keys, but we need to deal with their ephemeral existence on the scheduler.  The same applies to Locks/Queues/Variables and so on, which is a cause for complexity

File: distributed/scheduler.py (hunk 6 above)
Comment (mrocklin) [reply distributed/scheduler.py:4750]: There may be a performance reason, but in general I agree that we might want to remove stringification.  I would propose to handle that in follow-up work though if we do that.

File: distributed/client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rjzamora) [reply distributed/client.py:199]: Ahh - That makes a lot of sense. Thanks for clarifying!

File: distributed/scheduler.py (hunk 6 above)
Comment (rjzamora) [reply distributed/scheduler.py:4750]: I see. If stringification isn't an actual barrier here, then I completely agree with addressing that separately.

**Comment by mrocklin** [comment]
OK, my goal here is to get this done by the first week of May.  (a few of us plan to meet in Chicago and I'd like to work with folks on HLGs there).  (@rjzamora are you free?)

To get this done, there are a few things to do:

1. Still some cleanup around annotations (I'll do this next week)
2. Handling serialization of objects in the graph that aren't normally picklable, in particular H5Py and NetCDF objects.  If anyone here can help with custom Pickle serializers that would be really helpful.  (@rjzamora does anyone on your team have time and experience with this?)

**Comment by mrocklin** [comment]
Note to future self, I also need to set client.cancel to wait until the scheduler sends back the cancelled response.  Future cancellation state is now owned entirely by the scheduler, rather than being split between the scheduler and client.

**Comment by rjzamora** [comment]
>(a few of us plan to meet in Chicago and I'd like to work with folks on HLGs there). (@rjzamora are you free?)

Nice! I *should* be around :)

> in particular H5Py and NetCDF objects. If anyone here can help with custom Pickle serializers that would be really helpful. (@rjzamora does anyone on your team have time and experience with this?)

@madsbk and @jakirkham may have some insight, but I'd also be interested to learn more about this (and ask around).

**Comment by mrocklin** [comment]
OK, I think that everything works now except for types that we used to handle with custom serialization, like h5py and netcdf.

File: distributed/scheduler.py [hunk 9]
```diff
@@ -3792,25 +3801,162 @@ async def add_nanny(self, comm):
...
+        if isinstance(workers, (str, Number)):
+            workers = [workers]
+        if isinstance(workers, (tuple, set)):
+            workers = list(workers)
+        if isinstance(workers, list):
+            restrictions = workers
+        elif workers is None:
+            restrictions = []
+        else:
+            raise TypeError("Workers must be a list or set of workers or None")
```
Comment (ian-r-rose) [inline distributed/scheduler.py:3869]: Nit: can also be `(str, Number)`

File: distributed/scheduler.py [hunk 10]
```diff
@@ -3792,25 +3801,162 @@ async def add_nanny(self, comm):
...
+
+        dsk = dict(graph_)
```
Comment (ian-r-rose) [inline distributed/scheduler.py:3871]: `dask.utils.ensure_dict` has some minor optimizations around HLG layers

File: distributed/scheduler.py [hunk 11]
```diff
@@ -3792,25 +3801,162 @@ async def add_nanny(self, comm):
...
+
+        pre_stringify = set(dsk)
+        dsk = {stringify(k): stringify(v, exclusive=graph_) for k, v in dsk.items()}
+
+        def process(x, keys=dsk, string_keys=pre_stringify):
+            if callable(x):
+                return {stringify(k): x(k) for k in keys}
+            elif isinstance(x, (int, dict, tuple, set, list)):
+                return {stringify(k): x for k in string_keys or map(stringify, keys)}
+            elif isinstance(x, dict):
```
Comment (ian-r-rose) [inline distributed/scheduler.py:3903]: `dict` is in the above list, so I don't think this branch will ever be hit

File: distributed/scheduler.py [hunk 12]
```diff
@@ -4644,10 +4653,99 @@ def update_graph_hlg(
...
+        if user_priority:
+            user_priority = process(user_priority)
+        if restrictions:
+            restrictions = process(restrictions)
+
+        for layer in graph.layers.values():
+            if layer.annotations and "retries" in layer.annotations:
+                retries = retries or {}
+                d = process(layer.annotations["retries"], keys=layer, string_keys=None)
+                retries.update(d)  # TODO: there is an implicit ordering here
```
[the remark this answers is not in the document]
Comment (ian-r-rose) [reply distributed/scheduler.py:3924]: Do you still consider these TODO? I think this ordering is fine (even better if it's documented somewhere)

File: distributed/scheduler.py [hunk 13]
```diff
@@ -3792,25 +3801,162 @@ async def add_nanny(self, comm):
...
+        if annotations:
+            annotations = process(annotations)
+        if retries:
+            retries = process(retries)
+        if resources:
+            resources = process(resources)
+        if user_priority:
+            user_priority = process(user_priority)
+        if restrictions:
+            _restrictions = process(restrictions)
```
Comment (ian-r-rose) [inline distributed/scheduler.py:3916]: For the others you've reassigned the original variable, but here you've created a new one. I'm probably missing something here, but I don't see if `restrictions` is used further down anywhere?

File: distributed/scheduler.py [hunk 14]
```diff
@@ -3792,25 +3801,162 @@ async def add_nanny(self, comm):
...
+        else:
+            _restrictions = {}
+
+        for layer in graph_.layers.values():
+            if layer.annotations and "retries" in layer.annotations:
```
Comment (ian-r-rose) [inline distributed/scheduler.py:3921]: Nit: we are checking if `layer.annotations` exists a lot in this block. We could check once at the top:
```python
if not layer.annotations:
    continue
````

File: distributed/event.py [hunk 15]
```diff
@@ -177,11 +177,14 @@ class Event:
...
     def __init__(self, name=None, client=None):
-        try:
-            self.client = client or Client.current()
-        except ValueError:
-            # Initialise new client
-            self.client = get_worker().client
+        if client is None and _current_client.get() is not False:
+            try:
+                self.client = client or Client.current()
+            except ValueError:
```
[the remark this answers is not in the document]
Comment (ian-r-rose) [reply distributed/event.py:183]: As I'm reading further, I'm a bit uncomfortable with how often we see the this pattern, which is (a) accessing a private variable, and (b) overlaps in functionality with `Client.current()`. Any thoughts about retrofitting `Client.current()` to allow it to return `False` as well?

**Comment by ian-r-rose** [comment]
A thought I had today (which might be off base):

Previously, using GPU-enabled Dask clusters could be somewhat painful if you don't have a GPU on your local machine (cf https://github.com/rapidsai/cudf/issues/3661). That was one of the motivating use-cases of tools like `afar`. In this PR, the scheduler is doing much more graph manipulation, and in particular it is serializing tasks for workers. Does this mean that the scheduler would *also* need to have a GPU in those cases?

**Comment by jakirkham** [comment]
That's already the case as the Scheduler can deserialize things or be directly involved with serialization in some cases. This isn't desirable and something we would ideally fix, but it is likely no worse with this PR than is already the case. Though others more in the weeds here should feel free to correct this if I've missed something

**Comment by mrocklin** [comment]
> Does this mean that the scheduler would also need to have a GPU in those cases?

yes

**Comment by mrocklin** [comment]
Heads-up, my plan with this PR is to wait until stability to work is done and we have strong evidence that it's solid.  Then we can mint a new version that we're comfortable with and sit with that for a while.

Afterwards I'll work to merge this in and shake things up some more :)

**Comment by rjzamora** [comment]
Regarding [this comment in the HLG-roadmap issue](https://github.com/dask/dask/issues/9159#issuecomment-1179221598):  I'd like to help get this PR over the line, but my understanding of the client/scheduler code is quite limited compared to my understanding of dask/dask.

@mrocklin - Could you summarize what you expect the current state of this PR to be? Anything you know to be broken and/or missing?  I noticed that many tests fail when I use the `nuke-hlg` branches in both distributed and dask, and that there are some minor conflicts with `main`.  I will try to get these tests passing, but help/advice is very welcome.

**Comment by mrocklin** [comment]
@fjetter heads up, it looks like @rjzamora is becoming active here.  In principle I think that this is probably a good direction to go in (confidence 80% or so?) it's likely to cause mild havoc though.

**Comment by rjzamora** [comment]
>it looks like @rjzamora is becoming active here.

Yes - I started pushing on this a bit yesterday, and ~80% seems like a reasonable confidence level. On the dask/dask and HLG-development side, this is a huge win. However, we are making some clear trade-offs by *always* pickling HLGs (and there are probably issues we are not even considering yet).

One change that would probably make me feel more confident is if we preserved the old HLG-packing code path for the special case of materialized layers, and provided an option to materialize layers before shipping the hlg. We would still want to remove the dask_distributed_pack methods, but could preserve the pickle-free packing/unpacking logic somewhere. This would provide an escape hatch for users who are running into problems with pickle (maybe they have mismatched python environments, or no gpu on their scheduler, etc.).

**Comment by mrocklin** [comment]
> One change that would probably make me feel more confident is if we preserved the old HLG-packing code path for the special case of materialized layers, and provided an option to materialize layers before shipping the hlg. We would still want to remove the dask_distributed_pack methods, but could preserve the pickle-free packing/unpacking logic somewhere. This would provide an escape hatch for users who are running into problems with pickle (maybe they have mismatched python environments, or no gpu on their scheduler, etc.).

I'm inclined towards radical simplification here.  I don't think that we should do more half-measures here.  I think that we should burn things down.  (but others may disagree)

**Comment by rjzamora** [comment]
>I'm inclined towards radical simplification here. I don't think that we should do more half-measures here. I think that we should burn things down. (but others may disagree)

I'll submit something to your branch and see what you think. It seems that we can run the exact same code to materialize and process the HLG on either the client or the scheduler. So, I don't see a huge benefit in avoiding the option to run it on the client (yet).

**Comment by mrocklin** [comment]
> Could you summarize what you expect the current state of this PR to be? Anything you know to be broken and/or missing? I noticed that many tests fail when I use the nuke-hlg branches in both distributed and dask, and that there are some minor conflicts with main. I will try to get these tests passing, but help/advice is very welcome

I think that I had almost everything running smoothly.  There was some trickiness around Variables/Semaphores that I think I worked out almost completely but not entirely.

**Comment by fjetter** [comment]
So far, I only skimmed this PR.

At first glance, it looks like we're putting more emphasize on `Client.current` which makes me a bit nervous. The `Client.current` mechanism and specifically the default client mechanism is very complex and at times pretty unreliable, specifically in poorly isolated environments (many threads, asyncio, etc.)

This is a small thing, really, but I think we should use `Client.current(allow_default=False)` which actually returns None if there is no default client registered. I think this entire API should be cleaned up. 

Generally, if this is all about the serializability of Futures, maybe we should figure out how to (de-)serialize futures cleanly without touching client code? I might not have understood the problem, yet. Any pointers appreciated.

---


> > One change that would probably make me feel more confident is if we preserved the old HLG-packing code path for the special case of materialized layers, and provided an option to materialize layers before shipping the hlg. We would still want to remove the dask_distributed_pack methods, but could preserve the pickle-free packing/unpacking logic somewhere. This would provide an escape hatch for users who are running into problems with pickle (maybe they have mismatched python environments, or no gpu on their scheduler, etc.).
> 
> I'm inclined towards radical simplification here. I don't think that we should do more half-measures here. I think that we should burn things down. (but others may disagree)

How far would the "radical simplifications" go? Would this allow for simplifications in `distributed.protocol.serialized`, e.g. can we get rid of `Serialized`?

**Comment by mrocklin** [comment]
Quick update on this.  I'm still engaged, but have been pulled off for
stability work.  I hope that that work finishes up this week and then I'll
shift back to this and to shuffling work.

If anyone was interested, I'd love to also ge feedback on
https://github.com/dask/distributed/pull/6007
